lib/src/selection.dart

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// Where a selection start or end point may appear in some piece of text.
+abstract class Selection {
+  /// The chunk of text.
+  String get text;
+
+  /// The offset from the beginning of [text] where the selection starts, or
+  /// `null` if the selection does not start within this chunk.
+  int? get selectionStart => _selectionStart;
+  int? _selectionStart;
+
+  /// The offset from the beginning of [text] where the selection ends, or
+  /// `null` if the selection does not start within this chunk.
+  int? get selectionEnd => _selectionEnd;
+  int? _selectionEnd;
+
+  /// Sets [selectionStart] to be [start] characters into [text].
```

Comment:
Dart `length` and string indexes do not handle "characters" the way that authors think of them. Can we replace all mentions of "characters" in this class with "code units" or "UTF-16 code units"?